src/matrix.c

```diff
@@ -0,0 +1,130 @@
+#include "matrix.h"
+
+#include <math.h>
+
+/**
+ * @brief Converts a 2d coordinate in a matrix to an index in a 2d array
+ * @param row the row number
+ * @param col the column number
+ * @param size the number of columns in the matrix
+ */
+inline matrix_size coordToIndex(const matrix_size row, const matrix_size col, const matrix_size size) {
+    return row * size + col;
+}
+
+/**
+ * @brief Check if the given index is part of the diagonal
+ * @param index the index to check
+ * @param size the number of columns in the matrix
+ */
+inline matrix_size isDiagonal(const matrix_size index, const matrix_size size) {
+    return ((index / size) == (index % size));
+}
+
+/**
+ * @brief Multiplies 2 matrices together using this equation : output =
+ * firstMatrix x secondMatrix
+ * @param firstMatrix first matrix to multiply of size m * n
+ * @param secondMatrix second matrix to multiply of size n * p
+ * @param size array of size 3 containing m, n and p
+ * @param output matrix resulting of the matrix multiplication of the given
+ * matrices
+ * @param transposeFirstMatrix whether to transpose the first matrix when
+ * multiplying the matrices
+ */
+void matrixMultiply(const matrix_real_number* firstMatrix, const matrix_real_number* secondMatrix, const matrix_size size[3], matrix_real_number* output, const matrix_size transposeFirstMatrix) {
+    const matrix_size m = size[0];
+    const matrix_size n = size[1];
+    const matrix_size p = size[2];
+    for (matrix_size i = 0; i < m; ++i) {
+        for (matrix_size j = 0; j < p; ++j) {
+            matrix_real_number sum = 0.0;
+            for (matrix_size k = 0; k < n; ++k) {
+                sum += (transposeFirstMatrix ? firstMatrix[coordToIndex(k, i, n)] : firstMatrix[coordToIndex(i, k, n)]) * secondMatrix[coordToIndex(k, j, p)];
+            }
+            output[coordToIndex(i, j, p)] = sum;
+        }
+    }
+}
+
+/**
+ * @brief Creates the identity matrix (1 on the diagonal and 0 on other
+ * coordinates)
+ * @param size size of the matrix
+ * @param output output matrix
+ */
+void createIdentityMatrix(const matrix_size size, matrix_real_number* output) {
+    const matrix_size sizeSquared = size * size;
+    for (matrix_size i = 0; i < sizeSquared; ++i) {
+        if (isDiagonal(i, size)) {
+            output[i] = 1;
+        } else {
+            output[i] = 0;
+        }
```

Comment:
Why no use the tertiary operator here it seems like it would make the code more readable